File: backtracking/0051_n_queens.py
"""
Problem: N-Queens
LeetCode #: 51
Difficulty: Hard
Link: https://leetcode.com/problems/n-queens/

Approach: Backtracking placing one queen per row while tracking attacked columns, 
          positive diagonals (row + col), and negative diagonals (row - col).
Time Complexity: O(N!)
Space Complexity: O(N^2)
"""

from typing import List


class Solution:
    def solveNQueens(self, n: int) -> List[List[str]]:
        cols = set()
        posDiag = set()  # (r + c)
        negDiag = set()  # (r - c)

        res = []
        board = [["."] * n for _ in range(n)]

        def backtrack(r):
            if r == n:
                copy = ["".join(row) for row in board]
                res.append(copy)
                return

            for c in range(n):
                if c in cols or (r + c) in posDiag or (r - c) in negDiag:
                    continue

                cols.add(c)
                posDiag.add(r + c)
                negDiag.add(r - c)
                board[r][c] = "Q"

                backtrack(r + 1)

                cols.remove(c)
                posDiag.remove(r + c)
                negDiag.remove(r - c)
                board[r][c] = "."

        backtrack(0)
        return res


# --- Test ---
if __name__ == "__main__":
    sol = Solution()
    print("N = 4 Solutions:")
    for solution in sol.solveNQueens(4):
        for row in solution:
            print(row)
        print()

File: backtracking/0051_n_queens.java
/*
 * Problem: N-Queens
 * LeetCode #: 51
 * Difficulty: Hard
 * Link: https://leetcode.com/problems/n-queens/
 *
 * Approach: Backtracking placing one queen per row while tracking attacked columns,
 *           positive diagonals (r + c), and negative diagonals (r - c).
 * Time Complexity: O(N!)
 * Space Complexity: O(N^2)
 */

import java.util.*;

public class Solution {
    public List<List<String>> solveNQueens(int n) {
        List<List<String>> res = new ArrayList<>();
        char[][] board = new char[n][n];
        for (int i = 0; i < n; i++) {
            Arrays.fill(board[i], '.');
        }

        Set<Integer> cols = new HashSet<>();
        Set<Integer> posDiag = new HashSet<>();
        Set<Integer> negDiag = new HashSet<>();

        backtrack(0, n, board, res, cols, posDiag, negDiag);
        return res;
    }

    private void backtrack(int r, int n, char[][] board, List<List<String>> res,
                           Set<Integer> cols, Set<Integer> posDiag, Set<Integer> negDiag) {
        if (r == n) {
            List<String> copy = new ArrayList<>();
            for (char[] row : board) {
                copy.add(new String(row));
            }
            res.add(copy);
            return;
        }

        for (int c = 0; c < n; c++) {
            if (cols.contains(c) || posDiag.contains(r + c) || negDiag.contains(r - c)) {
                continue;
            }

            cols.add(c);
            posDiag.add(r + c);
            negDiag.add(r - c);
            board[r][c] = 'Q';

            backtrack(r + 1, n, board, res, cols, posDiag, negDiag);

            cols.remove(c);
            posDiag.remove(r + c);
            negDiag.remove(r - c);
            board[r][c] = '.';
        }
    }

    // --- Test ---
    public static void main(String[] args) {
        Solution sol = new Solution();
        List<List<String>> result = sol.solveNQueens(4);
        System.out.println("N = 4 Solutions Count: " + result.size());
        for (List<String> board : result) {
            for (String row : board) {
                System.out.println(row);
            }
            System.out.println();
        }
    }
}
